File: source/pypar_dist/source/network_timing.py
# Timing of MPI module for Python and estimation of latency and bandwidth
# 
# Send numerical array in a ring from processor 0 to 1 etc back to 0 
# Perform timings and compare different sending strategies 
#
# OMN, OCT 2001

import time, sys, pypar, numpy

# The send/recv routines
#
import pypar

#from mpiext import send_array, receive_array  #most direct methods
#Not available in final install - use bypass form for now

#--------------------------------------------------------------

method = 2   # Use 
             # 0: automatically allocated buffers
             # 1: user-supplied buffers
             # 2: Use bypass - let pypar use direct mpi_ext call
             # 3 (not there): direct call to mpiext (with buffers),
             #    only fractionally better than bypass
                  
vanilla = False  # Force use of vanilla format (1)
consistency_check = False # Check correctness


#--------------------------------------------------------------
# linfit
#
def linfit(x, y):
  """Fit a and b to the model y = ax + b. Return a,b,variance
  """
  
  Sx = Sy = SSoN = SxoN = norm = varest = 0.0
  N = len(x)
  assert len(y) == N, "x and y must have same length"
  
  for i in range(N):
    #print("x,y = %f, %f\n",x[i],y[i])
    Sx  = Sx + x[i]
    Sy  = Sy + y[i]
  
  SxoN = Sx/N
  
  a = 0.0 
  for i in range(N):
    t    = x[i] - SxoN
    SSoN = SSoN + t*t
    a    = a + t*y[i]


  a = a/SSoN            # a = (N Sxy - SxSy)/(NSxx - Sx^2) */
  b = (Sy - Sx*a)/N
  
  # Quality - variance estimate \sum_i r_i^2 /(m-n) 
  for i in range(N):  
    norm = norm + float(x[i])*x[i]
    res = y[i] - a*x[i] - b
    varest = varest + res*res

  varest = varest/norm/(N-2)
  return a, b, varest




#--------------------------------------------------------------
# Main program
#
MAXI  = 10         # Number of blocks 
MAXM  = 500000     # Largest block 
BLOCK = MAXM/MAXI  # Block size 

repeats = 10
msgid = 0
vanilla = 0 #Select vanilla mode (slower but general)

numprocs = pypar.size()
myid = pypar.rank()
processor_name = pypar.get_processor_name()

if myid == 0:
  # Main process - Create message, pass on, verify correctness and log timing
  #
  print "MAXM = %d, number of processors = %d" %(MAXM, numprocs)
  print "Measurements are repeated %d times for reliability" %repeats

if numprocs < 2:
  print "Program needs at least two processors - aborting\n"
  pypar.abort()
   
pypar.barrier() #Synchronize all before timing   
print "I am process %d on %s" %(myid,processor_name)


#Initialise data and timings
#

try:
  from numpy.random import uniform, seed
  seed(17)
  A = uniform(0.0,100.0,MAXM)
except:
  print 'problem with RandomArray'
  from numpy import ones, Float
  A = ones(MAXM).astype('f')
  
elsize = A.itemsize
#print elsize

noelem  = [0]*MAXI
bytes   = [0]*MAXI         
avgtime = [0.0]*MAXI         
mintime = [ 1000000.0]*MAXI      
maxtime = [-1000000.0]*MAXI            




if myid == 0:   
  # Determine timer overhead 
  cpuOH = 1.0;
  for k in range(repeats):   # Repeat to get reliable timings 
    t1 = pypar.time()
    t2 = pypar.time()
    if t2-t1 < cpuOH: cpuOH = t2-t1
    
  print "Timing overhead is %f seconds.\n" %cpuOH         

     
# Pass msg circularly   
for k in range(repeats):
  if myid == 0:
    print "Run %d of %d" %(k+1,repeats)
    
  for i in range(MAXI):
    m=BLOCK*i+1       
   
    noelem[i] = m
   
    pypar.barrier() # Synchronize 
   
    if myid == 0:
      #
      # Main process
      #
      t1 = pypar.time()

      if method==0:
        pypar.send(A[:m], destination=1, tag=msgid, vanilla=vanilla)
        C = pypar.receive(numprocs-1, tag=msgid, vanilla=vanilla)
      elif method == 1:  
        pypar.send(A[:m], use_buffer=True, destination=1,
                   tag=msgid, vanilla=vanilla)
        C = pypar.receive(numprocs-1, buffer=A[:m], tag=msgid, vanilla=vanilla)
      elif method==2:
        pypar.send(A[:m], use_buffer=True, destination=1,
                   tag=msgid, vanilla=vanilla, bypass=True)
        C = pypar.receive(numprocs-1, buffer=A[:m], tag=msgid,
                          vanilla=vanilla, bypass=True)
      else:
        raise 'Unknown method'
        #send_array(A[:m], 1, msgid)    
        #stat = receive_array(A[:m], numprocs-1, msgid)
        #C = A[:m]
        
      t2 = pypar.time() - t1 - cpuOH
      t2 = t2/numprocs
      avgtime[i] = avgtime[i] + t2
      if t2 < mintime[i]: mintime[i] = t2
      if t2 > maxtime[i]: maxtime[i] = t2

      # Uncomment to verify integrity of data
      # However, this may affect accuracy of timings for some reason.
      #
      if consistency_check:
        assert numpy.alltrue(C == A[:m])
    else:
      #
      # Parallel process - get msg and pass it on
      #

      if method==0:
        C = pypar.receive(myid-1, tag=msgid, vanilla=vanilla)
        pypar.send(A[:m], destination=(myid+1)%numprocs,
                   tag=msgid, vanilla=vanilla)
      elif method==1:  
        C = pypar.receive(myid-1, buffer=A[:m], tag=msgid, vanilla=vanilla)
        pypar.send(A[:m], use_buffer=True, destination=(myid+1)%numprocs,
                   tag=msgid, vanilla=vanilla)                        
      elif method==2:
        # Use pypar bypass
        C = pypar.receive(myid-1, buffer=A[:m], tag=msgid,
                          vanilla=vanilla, bypass=True)
        pypar.send(A[:m], use_buffer=True, destination=(myid+1)%numprocs,
                   tag=msgid, vanilla=vanilla, bypass=True)
      else:
        raise 'Unknown method'
        # Use direct mpiext call
        #stat = receive_array(A[:m], myid-1, msgid)                
        #send_array(A[:m], (myid+1)%numprocs, msgid)


# Output stats
#
if myid == 0:
  print "Bytes transferred   time (micro seconds)"
  print "                    min        avg        max "     
  print "----------------------------------------------"
     
  for i in range(MAXI):
    avgtime[i] = avgtime[i]/repeats*1.0e6 #Average micro seconds
    mintime[i] = mintime[i]*1.0e6         #Min micro seconds       
    maxtime[i] = maxtime[i]*1.0e6         #Min micro seconds              
          
    m = noelem[i]
    bytes[i] = elsize*noelem[i]       
      
    print "%10d    %10d %10d %10d" %(bytes[i], mintime[i], avgtime[i], maxtime[i]) 


  Tbw, Tlat, varest = linfit(bytes, mintime)
  print "\nLinear regression on best timings (t = t_l + t_b * bytes):\n",
  print "  t_b = %f\n  t_l = %f" %(Tbw,Tlat)
  print "  Estimated relative variance = %.9f\n" %varest
       
  print "Estimated bandwith (1/t_b):  %.3f Mb/s" %(1.0/Tbw)   
  print "Estimated latency:           %d micro s" %int(mintime[0]-bytes[0]*Tbw)  

   
pypar.finalize()




  
  

File: source/pypar_dist/source/network_timing.c
/*
  Estimate bandwidth and latency of a parallel computer using MPI.
  Ole Moller Nielsen - 1998
*/
	
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>


#define MAXI  10         /* Number of blocks */
#define MAXM  500000     /* Largest block */
#define BLOCK MAXM/MAXI  /* Block size */


double linfit(double* x, double* y, int N, double* a, double* b) 
{
  /* Given vectors y and x fit a and b to the model y = ax + b */
  
  double Sx=0, Sy=0, SxoN=0, SSoN=0, t=0;
  double res, varest=0, norm=0; 
  int i;
  
  for (i=0; i<N; i++)
  {
    /*printf("x,y = %f, %f\n",x[i],y[i]);*/
    Sx  = Sx + x[i];
    Sy  = Sy + y[i];
  }

  SxoN = Sx/N;
  
  *a = 0.0; 
  for (i=0; i<N; i++)
  {
    t    = x[i] - SxoN;
    SSoN = SSoN + t*t;
    *a   = *a + t*y[i];
  }

  *a = (*a)/SSoN;          /* a = (N Sxy - SxSy)/(NSxx - Sx^2) */
  *b = (Sy - Sx*(*a))/N;
  
  /* Quality - variance estimate \sum_i r_i^2 /(m-n) */
  for (i=0; i<N; i++)
  {
    norm = norm + x[i]*x[i];
    res = y[i] - (*a)*x[i] - (*b);
    varest = varest + res*res;
  } 
  varest = varest/norm/(N-2); 
  return(varest);
}

main(int argc, char **argv) 
{
   int repeats = 10, msgid = 0;
   int myid, procs;
   int i,j,k,m;

   double t1, t2, cpuOH; 
   double Tbw, Tlat;
   double varest;
      
   int noelem[MAXI];
   double bytes[MAXI];   
   double mintime[MAXI];   
   double maxtime[MAXI];      
   double avgtime[MAXI];         
   double A[MAXM]; 

   int  namelen;
   char processor_name[MPI_MAX_PROCESSOR_NAME];
   
   MPI_Status stat;
  
   
   /* Initialize */

   MPI_Init(&argc,&argv);
   MPI_Comm_size(MPI_COMM_WORLD,&procs);
   MPI_Comm_rank(MPI_COMM_WORLD,&myid);
   MPI_Get_processor_name(processor_name,&namelen);
   
   if (myid==0)
   {
     printf("MAXM = %d, number of processors = %d\n",MAXM,procs);         
     printf("Measurements are repeated %d times for reliability\n",repeats);
   } 
   
   if (procs < 2) {
     printf("Program needs at least two processors - aborting\n");
     MPI_Abort(MPI_COMM_WORLD,999);
   }
   
   MPI_Barrier(MPI_COMM_WORLD); /* Synchronize */   
   printf("I am process %d on %s\n",myid,processor_name);   
      
   for (j=0; j<MAXM; j++) 
   {
      A[j]=rand();
   }   
   for (i=0; i<MAXI; i++) 
   {
      avgtime[i] =  0;         
      mintime[i] =  1000000;      
      maxtime[i] = -1000000;            
   }
   
   /* Determine timer overhead */
   if (myid == 0) {   
     cpuOH = 1.0;
     for (k=0; k<repeats; k++)   /* Repeat to get reliable timings */
     { 
       t1 = MPI_Wtime();
       t2 = MPI_Wtime();
       if (t2-t1 < cpuOH) cpuOH = t2-t1;
     }  
     printf("Timing overhead is %f seconds\n\n", cpuOH);              
   }

   
        
   /* Pass msg circularly */
     
   for (k=0; k<repeats; k++) {
     if (myid == 0) {  
       printf("Run %d of %d\n", k+1, repeats);
     }  

     for (i=0; i<MAXI; i++) {
       /*m=BLOCK*(i+1);*/
       m=BLOCK*i+1;       
      
       noelem[i] = m;
      
       MPI_Barrier(MPI_COMM_WORLD); /* Synchronize */
      
       if (myid == 0) {
         t1=MPI_Wtime();
         MPI_Send(&A[0],m,MPI_DOUBLE,1,msgid,MPI_COMM_WORLD);
         MPI_Recv(&A[0],m,MPI_DOUBLE,procs-1,msgid,MPI_COMM_WORLD,&stat);
         t2=MPI_Wtime() - t1 - cpuOH;
	 t2 = t2/procs;
	 avgtime[i] = avgtime[i] + t2;
         if (t2 < mintime[i]) mintime[i] = t2;
         if (t2 > maxtime[i]) maxtime[i] = t2;	 
       } else {
         MPI_Recv(&A[0],m,MPI_DOUBLE,myid-1,msgid,MPI_COMM_WORLD,&stat);
         MPI_Send(&A[0],m,MPI_DOUBLE,(myid+1)%procs,msgid,MPI_COMM_WORLD);
       }
     } 
   }

   if (myid == 0) {
     printf("Bytes transferred   time (micro seconds)\n");
     printf("                    min        avg        max \n");     
     printf("----------------------------------------------\n");      
        
     for (i=0; i<MAXI; i++) {
       avgtime[i] = avgtime[i]/repeats*1.0e6; /*Average micro seconds*/
       mintime[i] = mintime[i]*1.0e6;         /*Min micro seconds*/       
       maxtime[i] = maxtime[i]*1.0e6;         /*Min micro seconds*/              
             
       m = noelem[i];
       bytes[i] = (double) 8*noelem[i];       
	 
       /* printf("m=%d, time(min)=%lf, time(avg)=%lf, time(max)=%lf\n",
	       m,mintime[i],avgtime[i],maxtime[i]); */
       printf("%10d    %10d %10d %10d\n",
	 (int) bytes[i], (int) mintime[i], (int) avgtime[i], (int)maxtime[i]); 
     }		 
   
     varest=linfit(bytes, mintime, MAXI, &Tbw, &Tlat);
     printf("\nLinear regression on best timings (t = t_l + t_b * bytes):\n");

     printf("  t_b = %f\n  t_l = %f\n", Tbw, Tlat);
     printf("  Estimated relative variance = %.9f\n\n",varest);     

     printf("Estimated bandwith (1/t_b):  %.3f Mb/s\n", (1.0/Tbw));         
     printf("Estimated latency:           %d micro s\n", 
             (int) (mintime[0] - (float) bytes[0]* (float)Tbw));         
     
   }

   MPI_Finalize();
}

